ticketutil/servicenow.py

```diff
@@ -0,0 +1,321 @@
+        :param ticked_id: ticked number, eg. PNT1234567
+        """
+        self.ticketing_tool = 'ServiceNow'
+        self.sys_id = None
+
+        # The auth param should be of the form (<username>, <password>) for
+        # HTTP Basic authentication.
+        self.url = url
+        self.auth = auth
+        self.table = project
+        self.rest_url = '{0}/api/now/v1/table/{1}'.format(self.url, self.table)
+        self.auth_url = self.rest_url
+
+        # parent creates request session, ticket_id beforehand causes failure
+        super(ServiceNowTicket, self).__init__(project, ticket_id=None)
+        if ticket_id:
+            super(ServiceNowTicket, self).set_ticket_id(ticket_id)
+
+        # OK for GET, but will it blend? Accept part might be needed later
+        self.s.headers.update({'Content-Type': 'application/json'})
+
+    def process_response(self, response):
+        """
+        Helper function to process Requests response or raise exception
+
+        :param response: response of the Requests query
+        """
+        try:
+            if (response.status_code != 201
+                and response.status_code != 200
+                ):
+                raise
+            result = response.json()['result']
+            return result
+        except:
+            logging.error('Headers: {0}'.format(response.headers))
+            logging.error('Error Response: {0}'.format(response.json()))
+
+    def api_get(self, url):
+        """
+        API GET wrapper
+
+        :param url: url containing REST API query
+        """
+        self.s.headers.update({'Content-Type': 'application/json'})
+        response = self.s.get(url)
+        logging.debug('GET request Status Code: {0}'
+                      .format(response.status_code))
+        return self.process_response(response)
+
+    def api_post(self, url, data):
+        """
+        API POST wrapper
+
+        :param url: url containing REST API query
+        :param text: text to be inserted into table record
+        """
+        self.s.headers.update({'Content-Type': 'application/json',
+                               'Accept': 'application/json'})
+        response = self.s.post(url, data=data)
+        logging.debug('POST request Status Code: {0}'
+                      .format(response.status_code))
+        return self.process_response(response)
+
+    def api_put(self, url, data):
+        """
+        API PUT wrapper
+        Updates table content.
+
+        :param url: url containing REST API query
+        :param text: text to be inserted into table record
+        """
+        self.s.headers.update({'Content-Type': 'application/json',
+                               'Accept': 'application/json'})
+        response = self.s.put(url, data=data)
+        logging.debug('PUT request Status Code: {0}'
+                      .format(response.status_code))
+        return self.process_response(response)
+
+    def get_sys_id(self):
```

Comment:
Actually they are both ticket ID. There are two related fields in the ServiceNow ticket:
1. 'sys_id' which contains hashcode that identifies ticket in the URL 
2. 'number' which contains something like 'PNT0123457' and can be seen as ticket ID in the WebUI.
So this method is needed when you generate ticket URL from the 'number'.